extensions/amp-user-notification/0.1/amp-user-notification.js

```diff
@@ -0,0 +1,423 @@
+import {viewerFor} from '../../../src/viewer';
+import {onDocumentReady} from '../../../src/document-state';
+import {xhrFor} from '../../../src/xhr';
+
+
+const TAG_ = 'AmpUserNotification';
+
+/**
+ * @export
+ * @typedef {{
+ *   elementId: string,
+ *   ampUserId: string
+ * }}
+ */
+let PostRequestMetadata;
+
+/**
+ * @export
+ * @typedef {{
+ *   showNotification: boolean
+ * }}
+ */
+let PostResponseMetadata;
+
+/**
+ * @typedef {{
+ *   promise: !Promise,
+ *   resolve: function(*)
+ * }}
+ */
+let UserNotificationDefer;
+
+
+class NotificationInterface {
+
+  /**
+   * Retrieves the id to associate to the Notification.
+   * @return {string}
+   */
+  getId() {}
+
+  /**
+   * Promise that is resolved with a boolean on wether this Notification
+   * should be shown or not.
+   * @return {!Promise<boolean>}
+   */
+  shouldShow() {}
+
+  /**
+   * Turns the Notification Component visible.
+   */
+  show() {}
+
+  /**
+   * Hides the Notification Component.
+   */
+  hide() {}
+
+  /**
+   * Returns a Promise that is only ever resolved if the current
+   * Notifiation has been dismissed by the user.
+   * @return {!Promise}
+   */
+  isDismissed() {}
+}
+
+/**
+ * Component class that handles a simple notification.
+ * @implements {NotificationInterface}
+ */
+export class AmpUserNotification extends AMP.BaseElement {
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return layout == Layout.NODISPLAY;
+  }
+
+  /**
+   * @private
+   */
+  isExperimentOn_() {
+
+  }
+
+  /** @override */
+  buildCallback() {
+
+    /** @private @const {!Window} */
+    this.win_ = this.getWin();
+
+    /** @private @const {!UserNotificationService} */
+    this.userNotification_ = getUserNotification_(this.win_);
+
+    this.elementId_ = assert(this.element.id, 'amp-user-notification should ' +
+        'have an id.');
+
+    /** @private {?string} */
+    this.ampUserId_ = null;
+
+    assert(this.element.hasAttribute('data-show-if-href'),
+        `"amp-user-notification" (${this.elementId_}) ` +
+        'should have "data-show-if-href" attribute.');
+    /** @private @const {string} */
+    this.showIfHref_ = assertHttpsUrl(
+        this.element.getAttribute('data-show-if-href'), this.element);
+
+    assert(this.element.hasAttribute('data-dismiss-href'),
+        `"amp-user-notification" (${this.elementId_}) ` +
+        'should have "data-dismiss-href" attribute.');
+    /** @private @const {string} */
+    this.dismissHref_ = assertHttpsUrl(
+        this.element.getAttribute('data-dismiss-href'), this.element);
+
+    /** @private {?Promise} */
+    this.getIfHrefPromise_ = null;
+
+    /** @private {?function(boolean)} */
+    this.dialogResolve_ = null;
+
+    /** @private @const {!Promise} */
+    this.layoutPromise_ = new Promise((resolve, reject) => {
+      this.layoutResolve_ = resolve;
+    });
+
+    /** @private @const {!Promise} */
+    this.dialogPromise_ = new Promise((resolve, reject) => {
```

Comment:
so we still need to initialize this promise earlier than that, since this is the same promise also passed into the cid service for the `persistentConsent` argument. no matter if we need to show the component or not we need a promise to expose to the service.